framework/src/bundle/CoreBundleContext.cpp

```diff
@@ -169,20 +169,32 @@ void CoreBundleContext::Uninit1()
   listeners.Clear();
   resolver.Clear();
 
+  // Do not hold the bundleThreads lock while calling
+  // BundleThread::Quit() because this will join the bundle
+  // thread which may need to acquire the bundleThreads lock
+  // itself.
+  //
+  // At this point, all bundles are stopped and all
+  // bundle context instances invalidated. So no new bundle
+  // threads can be created and it is sufficient to get
+  // the current list once and not check for new bundle
+  // threads again.
+  std::list<std::shared_ptr<BundleThread>> threads;
+  bundleThreads.Lock(), std::swap(threads, bundleThreads.value);
```

Comment:
This also means that when we swap out the `bundleThreads.value` list, we are already in a state where no code is executed any more that would add entries to it (all bundles are stopped). So this seems safe to me.